### ISSUE TITLE
[BUG] Animate Presence not working correctly with react-router-dom

### ISSUE THREAD

**Issue by MoSheikh** [framer/motion#650]
Hi,

I'm trying to get a setup going with which I can switch between routes and use AnimatePresence to fade the navigation screens in and out. However, the exit animation does not play at all when the component under the Route is unmounted.

https://codesandbox.io/s/condescending-sid-610y6?file=/src/App.tsx

Switching between the routes, you can see that the route screens fade in correctly, but immediately switch to the new route. The Animate Presence component should fade out the previous route before the new route fades in.

AnimatedPresence should stay mounted throughout the example, and the unique keys assigned to each child component are completely static so I don't believe the issue stems from there.

Help would be greatly appreciated.

**Comment by EugeneBugor** [comment]
Try to set `location={location}` for `Switch` component.

**Comment by jlevy-io** [comment]
Having exact same issue

**Comment by MoSheikh** [comment]
> Try to set `location={location}` for `Switch` component.

That fixed the issue for me! Thank you.

**Comment by MoSheikh** [comment]
> Try to set `location={location}` for `Switch` component.

That fixed the issue for me! Thank you.

**Comment by eltongonc** [comment]
@jlevy-io I've found out that the only way for this to work is by setting up your router in a specific way.  You must create a single `<Route />`  that renders your `<Switch location={location} key={location.pathname>` with all your routes.

[example](https://codesandbox.io/s/framer-motion-x-react-router-n7qhp)

**Comment by dinhnguyen92** [comment]
I also have this issue. @eltongonc's solution fixed it for me (thanks @eltongonc!), but it would be nice to not have to rely on a workaround that might not work in later versions of Framer Motion.

**Comment by dinhnguyen92** [comment]
After some experimentation, I found out a solution that does not require the use of an outer `<Route />` that renders the `Switch`. Basically, the key is to provide the `<Switch />` with both the `location` object and a `key`, just as @MoSheikh has pointed out. I guess the `key` is needed by `AnimatePresence` to track the components' presence in the DOM tree, which is why it has to be unique for each router path. Here's the solution:
```
<AnimatePresence exitBeforeEnter>
  <Switch location={location} key={location.pathname}>
    <Route path='/path'>
      {component}
    </Route>
  </Switch>
</AnimatePresence>
```
The key question is: how can I get the `location` object to pass into `<Switch />`? Perhaps @eltongonc came up with his workaround in order to solve this problem. However, I believe it is better to follow React Router documentation. Here's an excellent answer on the proper ways to obtain the `location` object:

https://stackoverflow.com/questions/37516919/react-router-getting-this-props-location-in-child-components

Basically, there are 2 main approaches to get the `location` object.

1. If your main app is a functional component:
```
import PropTypes from 'prop-types'
import { useLocation } from 'react-router'

function App() {
  const location = useLocation()
}
```

2. If your main app is React class:
```
import { withRouter } from 'react-router'

class App extends React.Component {
  static propTypes = {
    match: PropTypes.object.isRequired,
    location: PropTypes.object.isRequired,
    history: PropTypes.object.isRequired
  }

  render() {
    const { match, location, history } = this.props
  }
}

export default withRouter(App)
```

**Comment by singlexyz** [comment]
> 
> 
> @jlevy-io I've found out that the only way for this to work is by setting up your router in a specific way. You must create a single `<Route />` that renders your `<Switch location={location} key={location.pathname>` with all your routes.
> 
> [example](https://codesandbox.io/s/framer-motion-x-react-router-n7qhp)

Thanks a lot, this method save my world.
